src/solver/expressions/include/antares/solver/expressions/Registry.hxx

```diff
@@ -0,0 +1,52 @@
+/*
+** Copyright 2007-2024, RTE (https://www.rte-france.com)
+** See AUTHORS.txt
+** SPDX-License-Identifier: MPL-2.0
+** This file is part of Antares-Simulator,
+** Adequacy and Performance assessment for interconnected energy networks.
+**
+** Antares_Simulator is free software: you can redistribute it and/or modify
+** it under the terms of the Mozilla Public Licence 2.0 as published by
+** the Mozilla Foundation, either version 2 of the License, or
+** (at your option) any later version.
+**
+** Antares_Simulator is distributed in the hope that it will be useful,
+** but WITHOUT ANY WARRANTY; without even the implied warranty of
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+** Mozilla Public Licence 2.0 for more details.
+**
+** You should have received a copy of the Mozilla Public Licence 2.0
+** along with Antares_Simulator. If not, see <https://opensource.org/license/mpl-2-0/>.
+*/
+#pragma once
+
+#include <concepts>
+#include <memory>
+#include <mutex>
+#include <vector>
+
+namespace Antares::Solver
+{
+//  Template class to manage the memory allocation and registry for a base class
+template<class Base>
+class Registry
```

Comment:
Follows a talk with @flomnes.

This registry existence seems strange : it's both an object factory and a kind of garbage collector (it keeps memory of created object, in order to destroy them when it's destroyed).

How could we get rid of this strange object ?
The first idea that comes to mind : nodes destroying themselves, which means we would use smart pointer of nodes.
This idea was of course already considered.
The problem was : for node types that contain children (**AddNode** for instance), using smart pointers on nodes would cause complex notations defining constructors and data members.
A workaround would be to define a type alias : 
```C++
using NodePtr = std::shared_ptr<Node>;
``` 
Sorry, not a very original idea, but simple and more intuitive than the registry, in my opinion.